### PR TITLE
Speed up `mast.Observations.get_cloud_uris()`

### PR DESCRIPTION
**Pull request by barentsen**
When using `astroquery.mast`, it is currently very slow to download multiple data products in cloud-enabled mode.

The main reason for the poor performance is that the `Observations.get_cloud_uris()` helper method is very slow at retrieving AWS S3 URIs.  For example, the following example requires >2 minutes to obtain URIs for 186 data products:

Screen Shot 2021-09-07 at 5 44 33 PM

This is unfortunate because slow data access makes the cloud-enabled mode less attractive.

The main reason `Observations.get_cloud_uris()` is slow is because every data products triggers a separate HTTP request to MAST's `path_lookup` API via the `astroquery.mast.utils.mast_relative_path()` helper function.

This PR proposes to modify these functions such that MAST's `path_lookup` API is called in chunks of 50 products at once, allowing `Observations.get_cloud_uris()` to execute much faster.  For example, the demo above is ~7x faster when using this PR branch:

Screen Shot 2021-09-07 at 6 02 26 PM

(Note: the remaining 20s used by `get_cloud_uris()` is spent polling every cloud URI to verify that it exists before returning it to the user, because not every MAST product is mirrored in the cloud.  This will be harder to speed up, but fortunately this may be fast when running in the cloud.)

### CODE REVIEW

**Comment by bsipocz** [comment]
This now needs and rebase. Also, @jaymedina could you assign it to the person who are supposed to review this, or cc them on the PR? Thanks!

**Comment by jaymedina** [comment]
Hi @bsipocz I'll review this in conjunction with @tomdonaldson (I think he started a review already?)

File: astroquery/mast/cloud.py
Comment (jaymedina) [inline astroquery/mast/cloud.py:139]: This will break for GALEX data retrieval from the cloud, unfortunately. See here for how I tweaked this output in a recent PR following the new GALEX availability on the cloud: https://github.com/astropy/astroquery/blob/2b3d95498d0b6e73b9c3434ccc7d06cc50642f11/astroquery/mast/cloud.py#L119

Your two options would be to either: 
1. Make a tweak similar to this in `get_cloud_uri_list`
2. Make the tweak in `mast_relative_path` - this is probably the cleanest option. And then we can remove my tweak from `cloud.py`.

File: astroquery/mast/cloud.py
Comment (jaymedina) [inline astroquery/mast/cloud.py:146]: Should we raise an error here like how it was done in the original `get_cloud_uri`?:
`raise InvalidQueryError("Malformed data uri {}".format(data_product['dataURI']))`

**Review by jaymedina — CHANGES_REQUESTED** [review]
Looks solid, the main suggestion I have is what I mentioned about the GALEX relative path tweak that I discovered was necessary for GALEX on S3. Also added some minor comments, but in general this looks good. 

I'd say the first thing to do right now is rebase and rerun unit tests, to verify that GALEX indeed breaks without a tweak to the `mast_relative_path` output.

File: astroquery/mast/cloud.py
Comment (bsipocz) [reply astroquery/mast/cloud.py:139]: re this suggestion: in the cloud WG we're working on generalizing these methods and move them out from mast. So a tweak at the place of usage, or a generically usable kwarg would be preferable rather than a hard-wired conditional on mast specific strings.

**Comment by bsipocz** [comment]
@barentsen @jaymedina - what are the plans with this PR?

**Comment by bsipocz** [comment]
Ping @barentsen @jaymedina @tomdonaldson, do you have any plans for finishing up this PR? I expect it's reasonably close for being ready.

**Comment by bsipocz** [comment]
ping @jaymedina!

**Comment by jaymedina** [comment]
I'll take a look at this and make updates accordingly.

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py:144]: What I did was fold in the `path.strip()` logic out of `cloud.py` and into `util.py`'s `mast_relative_path` function. In other words, I essentially went with option `2.` listed above. And because this is a MAST-specific function in `utils`, it makes sense for MAST-specific tweaks to take place in there. Let me know if that works for you!

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py:151]: This comment is no longer necessary.

**Comment by jaymedina** [comment]
I've rebased, adapted the code to account for GALEX cloud products, updated error handling and unit tests, moved the MAST-specific string manipulation out of `cloud.py`, and squashed the commits into 4. I think this is ready for another review @bsipocz , thanks in advance.

File: astroquery/mast/cloud.py
Comment (bsipocz) [inline astroquery/mast/cloud.py:116]: Is this an InvalidQueryError or rather a `NoResultsWarning`? We haven't previously raised the error, but if you think the query has an error in it, rather than just there is no result, then we can raise it from now on (probably would require its own changelog entry)

File: astroquery/mast/cloud.py
Comment (bsipocz) [inline astroquery/mast/cloud.py:118]: Why only the first element, what is expected to be included in the rest?

File: astroquery/mast/cloud.py
Comment (bsipocz) [inline astroquery/mast/cloud.py:157]: I don't necessarily like the overwriting of variables here as when this has to be debugged it's a bit more difficult to distinguish between the `path` from the iteration vs the path here. Maybe use `s3_path`?

File: astroquery/mast/utils.py
Comment (bsipocz) [inline astroquery/mast/utils.py:185]: I don't see what's expected to be in `uri` here, could you explain it a bit, so I see why the second element?

File: astroquery/mast/tests/test_mast_remote.py
Comment (bsipocz) [inline astroquery/mast/tests/test_mast_remote.py:408]: do this with `pytest.mark.parametrize` instead of an internal loop

**Review by bsipocz — COMMENTED** [review]
Thanks for picking this up and pushing it through the finish line.

I have some minor comments, mostly asking for clarifications.

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py:119]: The output for `get_cloud_uri_list` is always a list, even when the output returns only 1 URI, so the sole element gets indexed out before being returned. I just committed some commentary above this line for clarity.

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py:116]: I chose to raise an error rather than a warning so that the output error message is more helpful. Here is what it looks like when I use a `NoResultsWarning`: 

image

The final `TypeError` is an indirect result of the empty uri list, so it's a vague error message. 

Raising a custom `InvalidQueryError` makes things more clear:

image

Comment (bsipocz) [reply astroquery/mast/cloud.py:116]: Maybe then a conditional is needed to return the empty result early, before going into the Typeerror. The point is basically to use the InvalidqueryError when there is an issue with the user input. Not having data, imo is not in this category.

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py:119]: There wouldn't be a case where a request for a single cloud URI would return more than one because of L171 in `util.py` of this branch, but I can raise an AssertionError here to ensure that there is only 1 element in the list, what do you think?

Comment (bsipocz) [reply astroquery/mast/cloud.py:119]: I'm really going into bikeshedding now, but if that's the case, why does it have to be a list to begin with?

Also, L173 is never reached? 

Maybe just a comment about this would be enough, no need for the assertion.

File: astroquery/mast/cloud.py
Comment (jaymedina) [reply astroquery/mast/cloud.py]: On second thought, a user will realize that their call to `Observations.get_cloud_uri(product)` returns an empty list before calling `Observations.download_file`, since they have to feed the uri in string form and by default the uri gets returned as a list even when it's 1 URI. So this error handling thing might be redundant.

**Comment by bsipocz** [comment]
Close/reopen to retrigger CI (as logs were long gone now)

**Comment by bsipocz** [comment]
Needs a rebase for RTD fix and some PEP8 fixes before final review.

**Comment by jaymedina** [comment]
Hi @bsipocz, I'm tagging the interim `astroquery.mast` maintainer, Julie Imig (@astrojimig). She will be taking over remaining PRs and will be the liaison for MAST-related issues.

**Comment by bsipocz** [comment]
@astrojimig -  I suppose the conflict is coming from the changes from your recently merged panstarss branch. Anyway, it would be nice to finally wrap this PR up and merge it. 

Could you try and rebase this one? I'm not sure you inherited the access to Geert's branch from Jenny.

**Comment by astrojimig** [comment]
@bsipocz Sure, I'll try to work on this next week!

**Comment by snbianco** [comment]
@bsipocz Working on this now! This might just be my own misunderstanding of GitHub, but do I need some sort of permissions to push to this branch? Whenever I try, I get a permission denied error. Is there an alternate method I should be using?

**Comment by bsipocz** [comment]
> @bsipocz Working on this now! This might just be my own misunderstanding of GitHub, but do I need some sort of permissions to push to this branch? Whenever I try, I get a permission denied error. Is there an alternate method I should be using?

You cannot push to someone else's fork by default, so you will need to take the applicable changes from here. I have reviewed this a couple of times, but things also have changed in the module, so I'm not certain which way is more efficient and useful (trying to keep these as is and rebase and fix from there, or look into the changes and apply the ones that are still relevant to your other PR (and then just rename your method and add fixes to it so old usage of get_cloud_uris() will work with it, too)

**Pull request closed without merging**
